pytential/collection.py

```diff
@@ -0,0 +1,427 @@
+__copyright__ = """
+Copyright (C) 2013 Andreas Kloeckner
+Copyright (C) 2018 Alexandru Fikl
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from typing import Dict, Hashable, Optional, Tuple, Union
+
+from pytential import sym
+from pytential.symbolic.execution import EvaluationMapperCSECacheKey
+
+from pytential.target import TargetBase
+from pytential.source import PotentialSource
+from pytential.qbx import QBXLayerPotentialSource
+from meshmode.discretization import Discretization
+
+__doc__ = """
+.. class:: GeometryLike
+
+    Types accepted by the :class:`GeometryCollection`.
+
+.. class:: DOFDescriptorLike
+
+    Types convertible to a :class:`~pytential.symbolic.primitives.DOFDescriptor`
+    by :func:`~pytential.symbolic.primitives.as_dofdesc`.
```

Comment:
Move this to `primitives.py`?

Reply:
`primitives` is already a bit long (2k lines or so). Move everything `DOFDesc`-y to `dof_desc.py`?

Reply:
Moved it in b6c3c7a and added some types.

There shouldn't be any functionality changes in there, but the diff sure looks scary :cry: